frame/im-online/src/lib.rs

```diff
@@ -231,7 +231,39 @@ pub trait WeightInfo {
 	fn validate_unsigned_and_then_heartbeat(k: u32, e: u32, ) -> Weight;
 }
 
-pub trait Trait: SendTransactionTypes<Call<Self>> + pallet_session::historical::Trait {
+/// Trait for retrieving the session info needed for online node inspection.
+///
+/// This trait is used for decouple the pallet-session dependency from im-online
+/// module so that the user of im-online & offences modules can pass any list of
+/// validators that are considered to be online in each session, particularly useful
+/// for the Substrate-based projects having their own staking implementation
+/// instead of using pallet-staking directly.
+pub trait SessionInterface<ValidatorId> {
```

Comment:
This trait looks good. What I meant with decoupling is to not depend on `pallet_session` in `im-online` pallet. I'd suggest we:
1. Put this trait into `primitives/session` (`sp-session`)
2. Put the below implementation into `pallet_session`, like so:
```
impl <T: crate::Trait> sp_session::SessionInterface<...> for crate::Module<T> {
  ...
}
```
So that it's implemented not for `Runtime`, but rather `type Session = pallet_session::Module<Runtime>;` That way you can write `type SessionInterface = Session;` in the implementation of `im-online` pallet.
3. Remove `pallet_session` dependency entirely from `im-online`. Perhaps this will require moving some other traits or types `IdentificationTuple`? to `sp-session`.

Reply:
- 1 & 2 has been done in d5a3b4f.
- For 3, I originally planned to put trait `VadalitorIdentification` and `IdentificationTuple` in `sp-session`, but `Parameter` is in `frame-support` which is not allowed to be introduced in primitives. Any suggestons?

https://github.com/paritytech/substrate/blob/b510b85e225fe3c88939fd1531867ce4ac8c4bdb/frame/session/src/historical/mod.rs#L47

We might also need to move `OneSessionHandler` into `sp-session` if we want to remove `pallet-session` from `pallet-im-online` entirely.

https://github.com/paritytech/substrate/blob/b510b85e225fe3c88939fd1531867ce4ac8c4bdb/frame/im-online/src/lib.rs#L621

Reply:
@tomusdrw Now it's enough to fix #7124 without any harm IMHO, but it's not much pretty clear to me how to remove `pallet-session` dep from `im-online` entirely, maybe we could create a new issue and refactor this in another PR later? ChainX is badly approaching the 2.0 upgrade, and we really need this patch.

Reply:
Ping @tomusdrw :-